_posts/2024-09-20-Detect-data-mutation-patterns-with-Debezium.adoc

```diff
@@ -0,0 +1,172 @@
+An important business metric to monitor would be the number of orders made.
+This number is very relevant to the business because it is directly correlated with the company’s earnings.
+A relevant drop in number of orders should alter the team responsible for that service as soon as possible.
+
+Normally, we would need to modify our application to instruct a https://prometheus.io/docs/concepts/metric_types/#counter[counter] metric that increments every time an order is correctly placed and stored in the database.
+But what if we could extract this information just by monitoring the database through CDC?
+
+Let’s explore how Debezium offers this possibility.
+
+== How Debezium Exposes Metrics
+
+Debezium exposes three categories of metrics:
+
+Snapshot metrics:: Provide information about connector operation while performing a snapshot, such as RowsScanned, RemainingTableCount, TotalTableCount.
+Streaming metrics:: Provide information about connector operation when the connector is reading the database log, such as TotalNumberOfCreateEventsSeen, QueueRemainingCapacity, QueueTotalCapacity.
+Schema history metrics:: Provide information about the status of the connector’s schema history, such as ChangesApplied, ChangesRecovered, Status.
+
+All these metrics are exposed through JMX. You can read more in our https://debezium.io/documentation/reference/stable/operations/monitoring.html[documentation].
+
+=== Database Activity Metrics
+
+Starting from the 3.0.0.Final release (specifically 3.0.0.Beta1), the following new streaming metrics have been introduced:
+
+NumberOfCreateEventsSeen:: Counts the number of create events per table.
+NumberOfDeleteEventsSeen:: Counts the number of delete events per table.
+NumberOfUpdateEventsSeen:: Counts the number of update events per table.
+NumberOfTruncateEventsSeen:: Counts the number of truncate events per table.
+
+Since these metrics are experimental, you need to opt-in to enable them by setting the internal.advanced.metrics.enable property to true.
+
+== Demo: Order service monitoring dashboard
+
+After you exposed these metrics with Debezium you can easily create a dashboard with alerts to monitor an eventual drastic drop that is significant for you business.
+We'll set up our order service with its own PostgreSQL database, Debezium PostgreSQL connector on the Kafka Connect runtime and Prometheus and Grafana as our observability platform.
+
+For this demo you need to clone the code in the https://github.com/debezium/debezium-examples[Debezium examples repository] and go to the `db-activity-monitoring` directory.
+
+== Components overview
+
+=== Order service
+The order service is a https://quarkus.io/[Quarkus] application that, per default, stores 100 orders every 10s, this mean a rate of ~10 orders per second.
+To simulate a wrong deployment, if you set the `app.version` application properties with a value different from `1.0`, it will start to insert orders with a drop of 50% of standard insert rate.
```

Comment:
```suggestion
You can simulate a bad deployment by setting the `app.version` property in application.properties to anything other than `1.0`. 
This will force the application to perform order inserts at half the standard rate.
```